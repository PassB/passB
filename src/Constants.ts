export enum executionContext {
  background = 'background',
  popup = 'popup',
  options = 'options',
}
