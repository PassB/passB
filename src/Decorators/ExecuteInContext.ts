// tslint:disable:no-any

import 'reflect-metadata';

export type ExecutionContext = 'background' | 'popup' | 'options' | undefined;
const executionContexts = ['background', 'popup'];

function getActionName(constructor: Function, propertyKey: string): string {
  return `${constructor.name}.${propertyKey}`;
}

interface AsynchronousRequest {
  action: string;
  params: any[];
}

export const executeInCorrectContext = () =>
  <R, T extends (this: any, ...args: any[]) => Promise<R>>
    (
      target: object,
      propertyKey: string,
      descriptor: TypedPropertyDescriptor<T>,
    ): TypedPropertyDescriptor<T> => {

    const executionContext = Reflect.getMetadata('executionContext', target, propertyKey);
    if (!executionContexts.includes(executionContext)) {
      throw new Error(`using ${executeInCorrectContext.name} `
        + 'decorator, but not declaring @Reflect.metadata("executionContext", "...") correctly',
      );
    }

    const wrappedFunction = descriptor.value || (() => Promise.reject(void 0));

    return {
      value: (async function wrapper(this: any, ...params: any[]): Promise<R> {
        if (window.executionContext === executionContext) {
          // console.debug('executing synchonously %s.%s', target.constructor.name, propertyKey);
          return await wrappedFunction.apply(this, params);
        }

        // console.debug('executing asynchonously %s.%s', target.constructor.name, propertyKey);
        return await browser.runtime.sendMessage({
          action: getActionName(target.constructor, propertyKey),
          params: [...params],
        });
      }) as T,
    };
  };

export const AsynchronousCallable = () =>
  <T extends { new(...args: any[]): {} }>(constructor: T): T =>
    class extends constructor {
      public constructor(...args: any[]) {
        super(...args);

        for (let obj = this; !!obj; obj = Object.getPrototypeOf(obj)) { // tslint:disable-line:no-this-assignment
          for (const propertyKey of Object.getOwnPropertyNames(obj)) {
            const executionContext = Reflect.getMetadata('executionContext', this, propertyKey);
            if (executionContext && window.executionContext === executionContext) {

              browser.runtime.onMessage.addListener(
                (request: AsynchronousRequest, sender: object, sendResponse: Function) => {
                  if (request.action !== getActionName(constructor, propertyKey)) {
                    return false;
                  }

                  const wrappedFn: ((...args: any[]) => Promise<any>) = (this as any)[propertyKey];
                  wrappedFn.apply(this, request.params).then(sendResponse);
                  return true;
                },
              );

            }
          }
        }
      }
    };

export const AsynchronousCallableServiceFactory =
  <X, T extends { new(...args: any[]): X }>(type: T): ((...args: any[]) => X) =>
    ((...args: any[]): X => (new (AsynchronousCallable()(type))(...args)));
