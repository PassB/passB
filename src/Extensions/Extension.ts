export interface ActionMetaData {
    [key: string]: string;
}

export interface ListEntry {
    label: string;
    actions: string[];
    metadata: ActionMetaData;
}

export interface Options {
}

export type RegisterEntryCallback = (entry: ListEntry) => void;

export abstract class Extension {
    public abstract readonly name: string;
    public abstract readonly actions: string[];
    protected readonly options: Options;

    public constructor(options: Options) {
        this.options = options;
    }

    public abstract initializeList(registerEntryCallback: RegisterEntryCallback): void;

    public abstract executeAction(action: string, entry: ListEntry): void;
}
