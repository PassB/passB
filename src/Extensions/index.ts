export {Extension, ExtensionTag, EntryActions, ExecutionOptions, RegisterEntryCallback} from './Extension';

import './PassExtension';
import './QRCodeExtension';
