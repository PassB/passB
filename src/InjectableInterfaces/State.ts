import {Store} from 'redux';
import {Persistor} from 'redux-persist';
import {StoreContents} from '../State/State';

export interface State {
  hydrated: Promise<void>;

  getPersistor(): Persistor;

  getStore(): Store<StoreContents>;
}
