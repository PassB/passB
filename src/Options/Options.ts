import {PassExtension} from "Extensions/PassExtension/PassExtension";
import {PassB} from "PassB";
import {OptionsReceiverInterface} from "./OptionsReceiver";

import deepExtend = require("deep-extend");

interface OptionsList {
  [extensionName: string]: {};
}

export interface OptionsData {
  enabledExtensions: string[];
  extensionsOptions: OptionsList;
  selectedMatcher: string;
  matchers: OptionsList;
  selectedFileFormat: string;
  fileFormats: OptionsList;
  selectedFiller: string;
  fillers: OptionsList;
}

function extractDefaultOptionsList(receivers: Array<OptionsReceiverInterface<{}>>): OptionsList {
  return receivers.reduce((list: OptionsList, receiver: OptionsReceiverInterface<{}>): OptionsList => ({
    ...list,
    [receiver.name]: receiver.defaultOptions,
  }), {} as OptionsList);
}

export class Options {
  private _options: OptionsData;

  public constructor(private passB: PassB) {
  }

  public async getOptions(): Promise<OptionsData> {
    if (!this._options) {
      this._options = deepExtend(
        this.getDefaultOptions(this.passB),
        (await browser.storage.local.get('options')) as OptionsData,
      );
    }
    return deepExtend({}, this._options);
  }

  public setOptions(newOptions: OptionsData): Promise<void> {
    this._options = newOptions;
    return browser.storage.local.set({options: newOptions});
  }

  private getDefaultOptions(passB: PassB): OptionsData {
    return {
      enabledExtensions: [PassExtension.name],
      extensionsOptions: extractDefaultOptionsList(passB.getAllExtensions()),
      selectedMatcher: passB.getAllMatchers()[0].constructor.name,
      matchers: extractDefaultOptionsList(passB.getAllMatchers()),
      selectedFileFormat: passB.getAllFileFormats()[0].constructor.name,
      fileFormats: extractDefaultOptionsList(passB.getAllFileFormats()),
      selectedFiller: passB.getAllFillers()[0].constructor.name,
      fillers: extractDefaultOptionsList(passB.getAllFillers()),
    };
  }
}
