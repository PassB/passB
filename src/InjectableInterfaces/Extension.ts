import {RouteProps} from 'react-router';
import {ExtensionName} from 'State/Interfaces';
import {MapTypeAllowedData, TypedMap} from 'State/Types/TypedMap';
import {OptionsPanelType} from './OptionsPanel';

export interface Extension<OptionType extends MapTypeAllowedData<OptionType>> {
  readonly routes: RouteProps[];
  readonly OptionsPanel?: OptionsPanelType<OptionType>;
  readonly actions: string[];
  readonly name: ExtensionName;
  readonly defaultOptions: TypedMap<OptionType>;
  initializeList(): Promise<void>;
  getLabelForAction(action: string): string;
  executeAction(action: string, entry: string, options: ExecutionOptions): void;

}

export interface ExecutionOptions {
  navigateTo: (newUrl: string, state: {}) => void;
}
