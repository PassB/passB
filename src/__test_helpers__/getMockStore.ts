import {combineReducers, createStore, Store} from 'redux';
import {StoreContents} from 'InjectableInterfaces/State';
import {initialState as initialOptionsState, reducer as OptionsReducer} from 'State/Options/Reducers';
import {initialState as initialPassEntriesState, reducer as PassEntryReducer} from 'State/PassEntries/Reducers';

const reducer = combineReducers<StoreContents>({
  options: OptionsReducer,
  passEntries: PassEntryReducer,
});

export const getMockStore = (): Store<StoreContents> => createStore<StoreContents>(
  reducer,
  {
    options: initialOptionsState,
    passEntries: initialPassEntriesState,
  },
);
