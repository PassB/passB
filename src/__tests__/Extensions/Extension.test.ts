// tslint:disable:no-unused-expression max-classes-per-file
import {Store} from 'redux';
import {Container, Service} from 'typedi';
import {getMockStore} from '__test_helpers__/getMockStore';
import {executionContext} from 'Constants';
import {setExecutionContext} from 'Decorators/ExecuteInContext';
import {ExecutionOptions, Extension} from 'Extensions/Extension';
import {OptionsPanelType} from 'Options/OptionsReceiver';
import {setExtensionDefaultOptions, setExtensionOptions} from 'State/Options/Actions';
import {State, StoreContents} from 'State/State';
import {createTypedMap} from 'State/Types/TypedMap';

@Service()
class TestExtension extends Extension<{}> {
  public OptionsPanel: OptionsPanelType<{}>;
  public actions: string[];

  public constructor() {
    super('testExtension', createTypedMap({defaultOption1: 'value1'}));
  }

  public initializeList(): Promise<void> {
    return Promise.resolve();
  }

  public getLabelForAction(action: string): string {
    return '';
  }

  public executeAction(action: string, entry: string, options: ExecutionOptions): void {
    return;
  }
}

describe('Extension base class', () => {
  let store: Store<StoreContents>;
  let state: { getStore: jest.Mock<Store<StoreContents>> };

  beforeAll(() => {
    state = {getStore: jest.fn((): Store<StoreContents> => store)};
  });

  beforeEach(() => {
    state.getStore = jest.fn((): Store<StoreContents> => store);
    store = getMockStore();
    Container.set(State, state);
    Container.remove(TestExtension);
  });

  describe('constructor', () => {
    it('should dispatch defaultOptions in background context', () => {
      setExecutionContext(executionContext.background);
      const spy = jest.spyOn(store, 'dispatch');

      Container.get(TestExtension);

      expect(spy.mock.calls).toMatchSnapshot();
    });

    Object.values(executionContext).filter((context: string) => context !== executionContext.background).forEach((context: string) => {
      it(`should not dispatch defaultOptions in context "${context}"`, () => {
        setExecutionContext(context);
        const spy = jest.spyOn(store, 'dispatch');

        Container.get(TestExtension);

        expect(spy).not.toHaveBeenCalled();
      });
    });

  });

  describe('onOptionsUpdate method', () => {
    it('should be triggered when options change', () => {
      const triggered = jest.fn();

      const extension = Container.get(class extends TestExtension {
        public onOptionsUpdate(): void {
          triggered(arguments);
        }
      });
      expect(triggered).not.toHaveBeenCalled();

      store.dispatch(setExtensionOptions({extensionName: extension.name, options: createTypedMap({foo: 'bar'})}));

      expect(triggered).toHaveBeenCalled();
      expect(triggered.mock.calls).toMatchSnapshot();

    });
  });
  describe('setEntries method', () => {
    it('should dispatch setEntries Action', () => {
      const extension = Container.get(TestExtension);
      const spy = jest.spyOn(store, 'dispatch');

      (extension as any) // tslint:disable-line:no-any
        .setEntries([{
          fullPath: 'foo/bar',
          action: 'test',
        }]);

      expect(spy.mock.calls).toMatchSnapshot();
    });
  });

  describe('options property', () => {
    it('should contain default options', () => {
      const defaultOptions = createTypedMap({foo: 'bar'});

      store = getMockStore();
      store.dispatch(setExtensionDefaultOptions({extensionName: 'testExtension', options: defaultOptions}));

      const extension = Container.get(TestExtension);

      expect(
        (extension as any).options, // tslint:disable-line:no-any
      ).toEqualImmutable(defaultOptions);
    });

    it('should contain a merge of default and explicitly set options', () => {
      const defaultOptions = createTypedMap({foo: 'bar'});
      const explicitOptions = createTypedMap({snafu: 'baz'});

      store = getMockStore();
      store.dispatch(setExtensionDefaultOptions({extensionName: 'testExtension', options: defaultOptions}));
      store.dispatch(setExtensionOptions({extensionName: 'testExtension', options: explicitOptions}));

      const extension = Container.get(TestExtension);

      expect(
        (extension as any).options, // tslint:disable-line:no-any
      ).toEqualImmutable(defaultOptions.merge(explicitOptions));
    });
  });
});
