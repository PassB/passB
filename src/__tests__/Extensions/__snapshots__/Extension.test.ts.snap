// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension base class constructor should dispatch defaultOptions in background context 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "extensionName": "testExtension",
        "options": Immutable.Map {
          "defaultOption1": "value1",
        },
      },
      "type": "OPTIONS/SET_EXTENSION_DEFAULT_OPTIONS",
    },
  ],
]
`;

exports[`Extension base class onOptionsUpdate method should be triggered when options change 1`] = `
Array [
  Array [
    Arguments [
      Immutable.Map {},
    ],
  ],
]
`;

exports[`Extension base class setEntries method should dispatch setEntries Action 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "entries": Array [
          Object {
            "action": "test",
            "fullPath": "foo/bar",
          },
        ],
        "extensionName": "testExtension",
      },
      "type": "PASS_ENTRIES/SET_ENTRIES",
    },
  ],
]
`;
